packages/arb-ts/src/lib/message/L2ToL1Message.ts

```diff
@@ -0,0 +1,374 @@
+  public readonly effectiveGasPrice: BigNumber
+  public readonly byzantium: boolean
+  public readonly type: number
+  public readonly status?: number
+
+  constructor(tx: TransactionReceipt) {
+    this.to = tx.to
+    this.from = tx.from
+    this.contractAddress = tx.contractAddress
+    this.transactionIndex = tx.transactionIndex
+    this.root = tx.root
+    this.gasUsed = tx.gasUsed
+    this.logsBloom = tx.logsBloom
+    this.blockHash = tx.blockHash
+    this.transactionHash = tx.transactionHash
+    this.logs = tx.logs
+    this.blockNumber = tx.blockNumber
+    this.confirmations = tx.confirmations
+    this.cumulativeGasUsed = tx.cumulativeGasUsed
+    this.effectiveGasPrice = tx.effectiveGasPrice
+    this.byzantium = tx.byzantium
+    this.type = tx.type
+    this.status = tx.status
+  }
+
+  public getL2ToL1Events() {
+    const iface = ArbSys__factory.createInterface()
+    const l2ToL1Event = iface.getEvent('L2ToL1Transaction')
+    const eventTopic = iface.getEventTopic(l2ToL1Event)
+
+    const logs = this.logs.filter(log => log.topics[0] === eventTopic)
+
+    return logs.map(
+      log => (iface.parseLog(log).args as unknown) as L2ToL1EventResult
+    )
+  }
+
+  private getOutboxAddr(network: Network, batchNumber: BigNumber) {
+    // CHRIS: add the old network to the networks object, and look it up here
+    // CHRIS: null check? this shouldn't be possible
+    // CHRIS: disable linting by just using the batchnumber here
+    batchNumber
+    return network.ethBridge!.outbox
+  }
+
+  public async getL2ToL1Messages<T extends SignerOrProvider>(
+    l1SignerOrProvider: T
+  ): Promise<L2ToL1MessageReaderOrWriter<T>[]>
+  public async getL2ToL1Messages<T extends SignerOrProvider>(
+    l1SignerOrProvider: T
+  ): Promise<L2ToL1MessageReader[] | L2ToL1MessageWriter[]> {
+    const provider = SignerOrProviderUtils.getProvider(l1SignerOrProvider)
+    if (!provider) throw new Error('Signer not connected to provider.')
+
+    const providerNetwork = await provider.getNetwork()
+    const arbNetwork = networks[providerNetwork.chainId]
+    return this.getL2ToL1Events().map(log => {
+      const outboxAddr = this.getOutboxAddr(arbNetwork, log.batchNumber)
+
+      return L2ToL1Message.fromBatchNumber(
+        l1SignerOrProvider,
+        outboxAddr,
+        log.batchNumber,
+        log.indexInBatch
+      )
+    })
+  }
+}
+
+/**
+ * Conditional type for Signer or Provider. If T is of type Provider
+ * then L2ToL1MessageReaderOrWriter<T> will be of type L2ToL1MessageReader.
+ * If T is of type Signer then L2ToL1MessageReaderOrWriter<T> will be of
+ * type L2ToL1MessageWriter.
+ */
+export type L2ToL1MessageReaderOrWriter<
+  T extends SignerOrProvider
+> = T extends Provider ? L2ToL1MessageReader : L2ToL1MessageWriter
+
+export class L2ToL1Message {
+  public static fromBatchNumber<T extends SignerOrProvider>(
+    l1SignerOrProvider: T,
+    outboxAddress: string,
+    batchNumber: BigNumber,
```

Comment:
Maybe we make a special `L2MessageId` interface that just combines these two (like we did with MessageID in L1ToL2Messages?